extractor/filesystem/language/python/uvlock/uvlock.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package uvlock extracts uv.lock files.
+package uvlock
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/BurntSushi/toml"
+	"github.com/google/osv-scalibr/extractor"
+	"github.com/google/osv-scalibr/extractor/filesystem"
+	"github.com/google/osv-scalibr/extractor/filesystem/language/python/internal/pypipurl"
+	"github.com/google/osv-scalibr/extractor/filesystem/osv"
+	"github.com/google/osv-scalibr/plugin"
+	"github.com/google/osv-scalibr/purl"
+)
+
+type uvLockPackageSource struct {
+	Virtual  string `toml:"virtual"`
+	Git      string `toml:"git"`
+	Registry string `toml:"registry"`
+}
+
+type uvLockPackage struct {
+	Name    string              `toml:"name"`
+	Version string              `toml:"version"`
+	Source  uvLockPackageSource `toml:"source"`
+
+	// uv stores "groups" as a table under "package" after all the packages, which due
+	// to how TOML works means it ends up being a property on the last package, even
+	// through in this context it's a global property rather than being per-package
+	Groups map[string][]uvOptionalDependency `toml:"optional-dependencies"`
+}
+
+type uvOptionalDependency struct {
+	Name string `toml:"name"`
+}
+type uvLockFile struct {
+	Version  int                               `toml:"version"`
```

Comment:
This is unused so let's not include

Reply:
I think there's some value in capturing it because it helps highlight that there is a concept of lockfile versions that we might have to one day account for - also, this is also present in the `poetrylock` extractor where it is also unused 😅 

I'm not actually fussed either way though so happy to remove if still if you'd prefer

Reply:
Let's keep the version here as we'll likely need it when the lockfile changes, and this is the one field that's guaranteed to be here even if the format changes. 

I'm wondering if we want to be even more defensive here and only extract lockfile versions that we expect, but I don't think we are doing that for other extractors, so I guess that is moot.